File: R/connectivity.R
library(signal)
library(stats)

compute_coherence <- function(sig1, sig2, fs) {
  spec <- mvspec(cbind(sig1, sig2), spans = c(3,3), plot = FALSE)
  f <- spec$freq * fs
  coh <- Mod(spec$fxx) / sqrt(Mod(spec$fxx) * Mod(spec$fyy))
  list(freq = f, coherence = coh)
}

compute_correlation_matrix <- function(data) {
  cor(data, use = "pairwise.complete.obs")
}

compute_phase_locking_value <- function(sig1, sig2) {
  phase1 <- Arg(fft(sig1))
  phase2 <- Arg(fft(sig2))
  plv <- abs(mean(exp(1i * (phase1 - phase2))))
  plv
}

File: python/connectivity.py
import numpy as np
import mne
from scipy.signal import coherence

def compute_coherence(raw, ch1, ch2, fmin=0.5, fmax=45):
    """Compute coherence between two channels in MNE Raw object."""
    data, times = raw[[ch1, ch2], :]
    f, Cxy = coherence(data[0], data[1], fs=raw.info['sfreq'])
    mask = (f >= fmin) & (f <= fmax)
    return f[mask], Cxy[mask]

def compute_correlation_matrix(raw):
    """Compute correlation matrix between all channels."""
    data = raw.get_data()
    return np.corrcoef(data)

def compute_phase_locking_value(raw, ch1, ch2):
    """Compute phase-locking value (PLV) between two channels."""
    data, _ = raw[[ch1, ch2], :]
    phase1 = np.angle(np.fft.fft(data[0]))
    phase2 = np.angle(np.fft.fft(data[1]))
    plv = np.abs(np.sum(np.exp(1j * (phase1 - phase2)))) / len(phase1)
    return plv
